tensorflow/compiler/mlir/xla/transforms/legalize_tf.cc

```diff
@@ -1299,6 +1300,193 @@ using ConvertDepthConv2DOp =
     ConvertConvOp<TF::DepthwiseConv2dNativeOp, /*num_spatial_dims=*/2,
                   /*depthwise_conv=*/true>;
 
+class ConvertPadOpStaticDynamic : public OpRewritePattern<TF::PadV2Op> {
+ public:
+  using OpRewritePattern::OpRewritePattern;
+  // Converts a tf.PadV2Op to mhlo.PadOp when input has static shape and
+  // padding values are const. Otherwise, converts the tf.PadV2Op to
+  // mhlo.DynamicPadOp.
```

Comment:
Sanjoy, your suggestion is more elegant. But this change costs too much (rewrite lots of Ops convert,  implement fold and  canonicalize func).

Reply:
I didn't quite get what is the cost here? 
Isn't it just making this function almost 1:1 and then moving this logic here to a canonicalization on `mhlo.DynamicPadOp`?

Reply:
(you can leave the canonicalization to a subsequent PR, but then don't add the code here in the first place please)

Reply:
In order to achieve the goals as Sanjoy suggest,  it's not just a matter of canonicalization but also rely on works of constant fold. See the example here:
`    %0 = mhlo.constant dense<[[1, 2]]> : tensor<1x2xi64>
    %1 = mhlo.constant dense<0> : tensor<1xi64>
    %2 = "mhlo.transpose"(%0) {permutation = dense<[1, 0]> : tensor<2xi64>} : (tensor<1x2xi64>) -> tensor<2x1xi64>
    %3 = "mhlo.reshape"(%2) : (tensor<2x1xi64>) -> tensor<2xi64>
    %4 = "mhlo.slice"(%3) {limit_indices = dense<1> : tensor<1xi64>, start_indices = dense<0> : tensor<1xi64>, strides = dense<1> : tensor<1xi64>} : (tensor<2xi64>) -> tensor<1xi64>
    %5 = "mhlo.slice"(%3) {limit_indices = dense<2> : tensor<1xi64>, start_indices = dense<1> : tensor<1xi64>, strides = dense<1> : tensor<1xi64>} : (tensor<2xi64>) -> tensor<1xi64>
    %6 = "mhlo.dynamic_pad"(%arg0, %arg1, %4, %5, %1) : (tensor<3xf32>, tensor<f32>, tensor<1xi64>, tensor<1xi64>, tensor<1xi64>) -> tensor<6xf32>`
Here the canonicalization of DynamicPad->Pad will not happen because the const fold for mhlo.slice is not supported yet.

Reply:
> (you can leave the canonicalization to a subsequent PR, but then don't add the code here in the first place please)

If you refer to the DynamicPad->Pad canonicalization pattern, it's a general optimization.

Reply:
> If you refer to the DynamicPad->Pad canonicalization pattern, it's a general optimization and not only for this goal.

Yes, that's what I referred to, and that's why I mentioned that you can leave it for later.
In the meantime, do we really need to add the special case in the lowering? Can we restrict ourself to implement the straightforward dynamic lowering only and not the static special case? We'll recover the performance with the folding+canonicalization later.

Reply:
> > If you refer to the DynamicPad->Pad canonicalization pattern, it's a general optimization and not only for this goal.
> 
> Yes, that's what I referred to, and that's why I mentioned that you can leave it for later.
> In the meantime, do we really need to add the special case in the lowering? Can we restrict ourself to implement the straightforward dynamic lowering only and not the static special case? We'll recover the performance with the folding+canonicalization later.

I would prefer to keep the static shape lowering patterns, especially if they were already there. The performance gap is big from our experience if the shape is actually known. They should be erased only after the the whole thing is done.

Reply:
It makes sense to me.  I removed static shape check in matchAndRewrite() here. Once folding and canonicalization func implemented and tf2mhlo pattern conversion for static shape removed(def in legalize_tf_patterns.td), TF Op would convert to dynamic mhlo pattern at first, and then recover perf by folding and canonicalization if it is static shape.